### ISSUE TITLE
Need hook for COD method availability 

### ISSUE THREAD

**Issue by menakas** [woocommerce/woocommerce#26074]
**Is your feature request related to a problem? Please describe.**
When a subscription product with free trial and no initial payment is purchased, the COD method becomes unavailable(https://d.pr/i/5TGDsK), if the "Enable for shipping methods" is set(https://d.pr/i/m33f5x).

This is because this logic on shipping packages [here](https://github.com/woocommerce/woocommerce/blob/master/includes/gateways/cod/class-wc-gateway-cod.php#L154-L166) returns without a filter.

**Describe the solution you'd like**
In this case, where we can modify `$needs_shipping` with `woocommerce_cart_needs_shipping`, if `enable_for_methods` is set, we are unable to get COD as an available payment method because there is no hook to modify the value.

**Comment by Konamiman** [comment]
Hi @menakas, thanks for notifying this.

I've been able to reproduce the issue and I've took a look at the involved code (thanks so much for pointing out the exact code snippet, by the way). Although a deeper investigation would be needed to be sure, at first sight it doesn't look like the solution is to add a hook; instead, this looks like a bug in the cart processing workflow.

The issue comes from the call to `get_canonical_package_rate_ids` in the code you highlighted. This is returning no results since `WC_Cart::get_shipping_packages` is returning an empty array, and this is due to a `woocommerce_cart_shipping_packages` filter in the subscriptions plugin that removes all the packages with free plugin, so that they don't count for the shipping cost.

So I'm not sure of how this should be fixed but it's definitely something that needs a fix.
